examples/platforms/emsk/flash.c

```diff
@@ -0,0 +1,175 @@
+#include "platform-emsk.h"
+
+/**
+ * EMSK HAS 128 Mbit (16 MB) SPI flash memory (Winbond W25Q128BV).
+ * The Winbond W25Q128BV array is organized into 65536 programmable pages of
+ * 256-byte each. Up to 256 bytes can be programmed at a time. Pages can be
+ * erased in groups of 16 (4KB sector erase).
+ */
+
+/**
+ * EMSK SPIflash memory map with using embARC
+ * FPGA image: from 0x0000 0000
+ * Secondary bootloader: from 0x0078 0000
+ * Available for OpenThread: from 0x00FF D000 to 0x00FF EFFF
+ */
+#define OPENTHREAD_FLASH_BASE       0x00ffd000
+#define OPENTHREAD_FLASH_SIZE       0x00002000
+
+/**
+ * Define SETTINGS_CONFIG_BASE_ADDRESS,
+ *    SETTINGS_CONFIG_PAGE_SIZE, and
+ *    SETTINGS_CONFIG_PAGE_NUM for EMSK
+ */
+#ifdef SETTINGS_CONFIG_BASE_ADDRESS
+#undef SETTINGS_CONFIG_BASE_ADDRESS
+#endif // SETTINGS_CONFIG_BASE_ADDRESS
+#define SETTINGS_CONFIG_BASE_ADDRESS    0x00ffe000
+
+#ifdef SETTINGS_CONFIG_PAGE_SIZE
+#undef SETTINGS_CONFIG_PAGE_SIZE
+#endif // SETTINGS_CONFIG_PAGE_SIZE
+#define SETTINGS_CONFIG_PAGE_SIZE   FLASH_SECTOR_SIZE
+
+#ifdef SETTINGS_CONFIG_PAGE_NUM
+#undef SETTINGS_CONFIG_PAGE_NUM
+#endif // SETTINGS_CONFIG_PAGE_NUM
+#define SETTINGS_CONFIG_PAGE_NUM    1
+
+ThreadError utilsFlashInit(void)
+{
+    flash_init();
+    return kThreadError_None;
+}
+
+uint32_t utilsFlashGetSize(void)
+{
+    return (uint32_t)FLASH_SECTOR_SIZE;
+}
+
+ThreadError utilsFlashErasePage(uint32_t aAddress)
+{
+    ThreadError error = kThreadError_None;
+    int32_t status;
+
+    VerifyOrExit((aAddress >= OPENTHREAD_FLASH_BASE) && (aAddress < (OPENTHREAD_FLASH_BASE + OPENTHREAD_FLASH_SIZE - 1)),
+            error = kThreadError_InvalidArgs);
+
+    /* Use 2 sectors in the implementation, cannot erase the address over the boundry */
+    status = flash_erase(aAddress, FLASH_SECTOR_SIZE);
+
+    VerifyOrExit(!(status == -1), error = kThreadError_Failed);
+
+exit:
+    return error;
+}
+
+ThreadError utilsFlashStatusWait(uint32_t aTimeout)
+{
+    ThreadError error = kThreadError_None;
+    uint32_t start = otPlatAlarmGetNow();
+    uint32_t busy = 1;
+    uint32_t status = 0x01;
+
+    while (busy && ((otPlatAlarmGetNow() - start) < aTimeout))
+    {
+        status = flash_read_status();
+        if (status & 0x01)
+        {
+            busy = 1;
+        }
+        else
+        {
+            busy = 0;
+        }
+    }
+
+    VerifyOrExit(!busy, error = kThreadError_Busy);
+
+exit:
+    return error;
+}
+
+uint32_t utilsFlashWrite(uint32_t aAddress, uint8_t *aData, uint32_t aSize)
+{
+    int32_t written_size = 0;
+    uint32_t size = 0;
+
+    VerifyOrExit((aAddress >= OPENTHREAD_FLASH_BASE) &&
+            ((aAddress + aSize) <= OPENTHREAD_FLASH_BASE + OPENTHREAD_FLASH_SIZE) &&
+            (!(aAddress & 3)) && (!(aSize & 3)), ;);
+
+    written_size = flash_write(aAddress, aSize, aData);
+    if (written_size < 0)
+    {
+        size = 0;
+    }
+    else
+    {
+        size = (uint32_t)written_size;
```

Comment:
Yes. The OpenThread core in `src` is implemented by C++. static_cast is efficient and safe for type conversion in C++. 
It is very good that OpenThead provide a C interface for platform implementation. So different hardware platform can support OpenThread stack smoothly. In technical point of view, it is possible to use C++ to implement hardware platform. It also means a lot of change in toolchain setting, code and build system side. EMSK BSP is based on C. And other hardware platforms in OpenThread also are implemented in pure C. 
I think it would be more acceptable if using a pure C implementation for hardware platform. Is it OK? : )